### ISSUE TITLE
node_cpu_hourly_cost value not reconciling with actual Azure Price Rates

### ISSUE THREAD

**Issue by danijam** [opencost/opencost#2640]
**Describe the bug**
We are trying to reconcile the data results from OpenCost for our AKS cluster. The results from our prometheus `/api/v1/query?query=node_cpu_hourly_cost` for the Azure VM SKU `Standard_DS2_v2` give us a value of `$0.058301/core/hour` That's a two core machine so the total hourly cost for the VM instance is `$0.116602/hour`. When we download our Azure Billing Account/Billing Profile Pricing Sheet and lookup this VM SKU the hourly price is `$0.1360/hour`. Meaning OpenCost is using a hourly price that is off by `$0.019398/hour` for nodes in our cluster using `Standard_DS2_v2` VMs.

We are a MCA Azure customer (Microsoft Customer Agreement) which uses standard pay-as-you-go rates with Offer Id ms-azr-0017g (https://azure.microsoft.com/en-gb/pricing/offers/ms-azr-0017g/)

I noticed that the logs from OpenCost is referring to `MS-AZR-0003p` (https://azure.microsoft.com/en-gb/pricing/offers/ms-azr-0003p/) as the durable offer id. That I believe refers to customers who contract with Azure under the general Microsoft Online Subscription Agreement . (https://azure.microsoft.com/en-gb/support/legal/subscription-agreement). This also using allegedly using standard PAYG rates.

I can't prove if the offer id difference is to blame, in theory I would not expect it to given they both use standard PAYG rates.

We are not a Enterprise Agreement Customer and so can't use the instructions here https://www.opencost.io/docs/configuration/azure#customer-specific-pricing

Would love some assistance to debug this further.

**To Reproduce**
1. Create a AKS cluster (no specific configuration required)
2. Install Prometheus (https://www.opencost.io/docs/installation/prometheus)
3. Install OpenCost (https://www.opencost.io/docs/installation/helm)
4. Query `query=node_cpu_hourly_cost` (https://www.opencost.io/docs/troubleshooting#verifying-pricing)
5. From Azure Portal - Navigate to `Cost Management + Billing`
6. Select Billing Account/Billing Scope where you AKS cluster is being billed
7. (If you have Billing Profiles go to the correct Billing Profile
8. From the Summary page select "Download Usage and Prices" from the "Shortcuts" box
9. Download price sheet and locate VM price rate
![image](https://github.com/opencost/opencost/assets/40212250/ad6fbf10-796c-4901-9490-ee02aa27de85)

**Expected behaviour**
I can reconcile the `node_cpu_hourly_cost` value to the Azure price sheet for a given VM SKU/Region

**Which version of OpenCost are you using?**
1.109.0 (Chart version 1.31.0)

**Additional context**
AKS - k8s version 1.28.3

**Comment by mattray** [comment]
I'm not familiar with the issue you're facing, but if @r2k1, @Sean-Holcomb or any of the other @opencost/opencost-maintainers are more familiar with Azure billing they might be able to weigh in.

**Comment by danijam** [comment]
@mattray are you are at kubecon? Happy to come say hi 👋

**Comment by mattray** [comment]
@danijam I am, happy to follow up in Slack or outside GitHub :) I'll be at the OpenCost kiosk Friday 12:30-2:30 in the Project Pavilion.

**Comment by raghavendra-rafay** [comment]
@danijam Did find the RCA of the issue, I am facing a similar issue for an Enterprise Agreement Azure Account. I am having a price difference between Opencost (node_total_hourly_cost) and Azure's Cost and Usage Report (unitPrice).

@Sean-Holcomb @mattray Please provide your inputs here. 

Opencost seems to query Price Sheet API from Consumption Service (https://learn.microsoft.com/en-us/rest/api/consumption/price-sheet?view=rest-consumption-2023-05-01). While Cost and Usage Report from Cost Management Service which has a different Price Sheet API (https://learn.microsoft.com/en-us/rest/api/cost-management/price-sheet?view=rest-cost-management-2023-11-01).

Should OpenCost use Cost Management's Price Sheet API to fetch the correct Unit Price?

**Comment by mattray** [comment]
I wonder if this is related to https://github.com/opencost/opencost/issues/2247

I'm not an Azure expert (paging @r2k1), but I'm under the impression that if we're still using the deprecated Azure Rate Card API we may not get the right pricing. @AjayTripathy or @Sean-Holcomb should probably weigh in on this too since there are several potentially related issues that we could close/fix:

https://github.com/opencost/opencost/issues/2601
https://github.com/opencost/opencost/issues/2714
https://github.com/opencost/opencost/issues/2737
https://github.com/opencost/opencost/issues/2826

**Comment by danijam** [comment]
@raghavendra-rafay No, we ended up using Kubecost (It still is using OpenCost under the hood). Incidently Kubecost was also presenting inaccurate pricing information, however Kubecost has a billing reconciliation capability which means that it can adjust pricing to make it accurate.

**Comment by armandocollazo-tomtom** [comment]
Hi everyone!
Is there any updates about this? We are using OpenCost v1.119.1 and just realized we have the exact same issue. We have an MCA too.  From the logs I see it is making a request to `https://management.azure.com/subscriptions/<subscription-id>/providers/Microsoft.Commerce/RateCard?$filter=OfferDurableId+eq+'MS-AZR-0003p'+and+Currency+eq+'USD'+and+Locale+eq+'en-US'+and+RegionInfo+eq+'US'&api-version=2015-06-01-preview`.

The region is also wrong but I believe that could be set in the configuration, right?
